packages/flutter/lib/src/foundation/persistent_hash_map.dart

```diff
@@ -0,0 +1,340 @@
+
+  final int occupiedIndices;
+  final List<Object?> keyValuePairs;
+
+  @override
+  _TrieNode copyWith(int bitIndex, Object key, int keyHash, Object? value) {
+    final int bit = 1 << _TrieNode.trieIndex(keyHash, bitIndex);
+    final int index = _compressedIndex(bit);
+
+    if ((occupiedIndices & bit) != 0) {
+      // Index is occupied.
+      final Object? keyOrNull = keyValuePairs[2 * index];
+      final Object? valueOrNode = keyValuePairs[2 * index + 1];
+
+      // Is this a (null, trieNode) pair?
+      if (identical(keyOrNull, null)) {
+        // ignore: cast_nullable_to_non_nullable
+        final _TrieNode newNode = (valueOrNode as _TrieNode).copyWith(
+            bitIndex + _TrieNode.hashBitsPerLevel, key, keyHash, value);
+        if (newNode == valueOrNode) {
+          return this;
+        }
+        return _CompressedNode(
+            occupiedIndices, _copy(keyValuePairs)..[2 * index + 1] = newNode);
+      }
+
+      if (key == keyOrNull) {
+        // Found key/value pair with a matching key. If values match
+        // then avoid doing anything otherwise copy and update.
+        return identical(value, valueOrNode)
+            ? this
+            : _CompressedNode(
+                occupiedIndices, _copy(keyValuePairs)..[2 * index + 1] = value);
+      }
+
+      // Two different keys at the same index, resolve collision.
+      final _TrieNode newNode = _resolveCollision(
+          bitIndex + _TrieNode.hashBitsPerLevel,
+          keyOrNull,
+          valueOrNode,
+          key,
+          keyHash,
+          value);
+      return _CompressedNode(
+          occupiedIndices,
+          _copy(keyValuePairs)
+            ..[2 * index] = null
+            ..[2 * index + 1] = newNode);
+    } else {
+      // Adding new key/value mapping.
+      final int occupiedCount = _bitCount(occupiedIndices);
+      if (occupiedCount >= 16) {
+        // Too many occupied: inflate compressed node into full node and
+        // update descendant at the corresponding index.
+        return _inflate(bitIndex)
+          ..descendants[_TrieNode.trieIndex(keyHash, bitIndex)] =
+              _CompressedNode.empty.copyWith(
+                  bitIndex + _TrieNode.hashBitsPerLevel, key, keyHash, value);
+      } else {
+        // Grow keyValuePairs by inserting key/value pair at the given
+        // index.
+        final int prefixLength = 2 * index;
+        final int totalLength = 2 * occupiedCount;
+        final List<Object?> newKeyValuePairs =
+            List<Object?>.filled(totalLength + 2, null);
+        for (int i = 0; i < prefixLength; i++) {
+          newKeyValuePairs[i] = keyValuePairs[i];
+        }
+        newKeyValuePairs[prefixLength] = key;
+        newKeyValuePairs[prefixLength + 1] = value;
+        for (int i = prefixLength, j = prefixLength + 2;
+            i < totalLength;
+            i++, j++) {
+          newKeyValuePairs[j] = keyValuePairs[i];
```

Comment:
nit: flutter style guide says not to use `i` or `j` iterator names - perhaps something like `keyValueIndex = prefixLength, newKeyValueIndex = prefixLength + 2` etc.